internal/persistence/postgres/options.go

```diff
@@ -0,0 +1,111 @@
+package postgres
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/kong/koko/internal/persistence"
+	"github.com/samber/lo"
+	"go.uber.org/zap"
+)
+
+var (
+	defaultSSLMode            = "verify-full"
+	unsupportedParamErrPrefix = "the '%s' parameter is unsupported, "
+	unsupportedParamErrs      = map[string]string{
+		"host":        unsupportedParamErrPrefix + " please instead use 'postgres.hostname'",
+		"port":        unsupportedParamErrPrefix + " please instead use 'postgres.port'",
+		"dbname":      unsupportedParamErrPrefix + " please instead use 'postgres.db_name'",
+		"user":        unsupportedParamErrPrefix + " please instead use 'postgres.user'",
+		"password":    unsupportedParamErrPrefix + " please instead use 'postgres.password'",
+		"sslrootcert": unsupportedParamErrPrefix + " please instead use 'postgres.ca_bundle_path'",
+	}
+)
+
+type Opts struct {
+	DBName           string
+	Hostname         string
+	ReadOnlyHostname string
+	Port             int
+	User             string
+	Password         string
+	EnableTLS        bool
+	CABundleFSPath   string
+	SQLOpen          persistence.SQLOpenFunc
+
+	// Parameters passed to the Postgres DB driver (pgx).
+	//
+	// This is here to allow the enablement of additional parameters, like `sslmode=verify-ca`.
+	// A subset of the connection parameters supported by libpq are supported by pgx. Additionally,
+	// pgx also lets you specify runtime parameters directly in the connection string.
+	//
+	// See: http://www.postgresql.org/docs/current/static/libpq-connect.html#LIBPQ-CONNSTRING for more
+	// information about connection parameters.
+	// See: https://www.postgresql.org/docs/current/runtime-config.html for more information about runtime parameters.
+	Params map[string]string
+}
+
+// Validate ensures the provided Postgres options are a valid configuration.
+func (opts *Opts) Validate() error {
+	for key := range opts.Params {
+		if msg, ok := unsupportedParamErrs[key]; ok {
+			return fmt.Errorf(msg, key)
+		}
+	}
+
+	return nil
+}
+
+func (opts *Opts) DSN(logger *zap.Logger) (string, error) {
+	var dsn string
+	sslmode := defaultSSLMode
+
+	if opts.Hostname != "" {
+		dsn += fmt.Sprintf("host=%s ", opts.Hostname)
+	}
+	if opts.Port != 0 {
+		dsn += fmt.Sprintf("port=%d ", opts.Port)
+	}
+	if opts.User != "" {
+		dsn += fmt.Sprintf("user=%s ", opts.User)
+	}
+	if opts.Password != "" {
+		dsn += fmt.Sprintf("password=%s ", opts.Password)
+	}
+	if opts.DBName != "" {
+		dsn += fmt.Sprintf("dbname=%s", opts.DBName)
+	}
+
+	if !opts.EnableTLS {
+		logger.Info("using non-TLS Postgres connection")
+		sslmode = "disable"
+	} else {
+		logger.Info("using TLS Postgres connection")
+		logger.Info("ca_bundle_fs_path:" + opts.CABundleFSPath)
+		if opts.CABundleFSPath == "" {
+			return "", fmt.Errorf("postgres connection requires TLS but ca_bundle_fs_path is empty")
+		}
```

Comment:
nit: You don't have to change anything here as it's existing code, and well, it's not directly related to this PR anyway.

But requiring this when TLS is enabled is weird, as it's an optional field:
https://github.com/jackc/pgx/blob/987de3874e0fc18b8c468d27cb9de9b2cd67540c/pgconn/config.go#L635-L638

Figured we might want to patch it up while we're at it, so figured to bring it up.

Reply:
good point. I'm for removing this now.

Reply:
done